lavalink/__init__.py

```diff
@@ -1,38 +1,84 @@
-__author__ = 'Devoxin'
-__license__ = 'MIT'
-__copyright__ = 'Copyright 2018 Devoxin'
-__version__ = '3.0.0'
-
-
-import logging
-import sys
-from .client import Client
-from .events import TrackStartEvent, TrackStuckEvent, TrackExceptionEvent, TrackEndEvent, QueueEndEvent
-from .models import BasePlayer, DefaultPlayer, AudioTrack, NoPreviousTrack, InvalidTrack
-from .node import Node
-from .nodemanager import NodeManager
-from .playermanager import PlayerManager
-from .utils import format_time
-from .websocket import WebSocket
-
-
-def enable_debug_logging():
-    """
-    Sets up a logger to stdout. This solely exists to make things easier for
-    end-users who want to debug issues with Lavalink.py.
-    """
-    log = logging.getLogger(__name__)
-
-    fmt = logging.Formatter(
-        '[%(asctime)s] [lavalink.py] [%(levelname)s] %(message)s',
-        datefmt="%H:%M:%S"
-    )
-
-    handler = logging.StreamHandler(sys.stdout)
-    handler.setFormatter(fmt)
-    log.addHandler(handler)
-
-    log.setLevel(logging.DEBUG)
+# flake8: noqa
+
+__title__ = 'Lavalink'
+__author__ = 'Devoxin'
+__license__ = 'MIT'
+__copyright__ = 'Copyright 2019 Devoxin'
+__version__ = '3.0.0'
+
+
+import logging
+import sys
+import inspect
+from .client import Client, _event_hooks
+from .events import Event, TrackStartEvent, TrackStuckEvent, TrackExceptionEvent, TrackEndEvent, QueueEndEvent, \
+    NodeConnectedEvent, NodeChangedEvent, NodeDisconnectedEvent, WebSocketClosedEvent
+from .exceptions import NodeException
+from .models import BasePlayer, DefaultPlayer, AudioTrack, NoPreviousTrack, InvalidTrack
+from .node import Node
+from .nodemanager import NodeManager
+from .playermanager import PlayerManager
+from .stats import Penalty, Stats
+from .utils import format_time, parse_time
+from .websocket import WebSocket
+
+
+def enable_debug_logging():
+    """
+    Sets up a logger to stdout. This solely exists to make things easier for
+    end-users who want to debug issues with Lavalink.py.
+    """
+    log = logging.getLogger('lavalink')
+
+    fmt = logging.Formatter(
+        '[%(asctime)s] [lavalink.py] [%(levelname)s] %(message)s',
+        datefmt="%H:%M:%S"
+    )
+
+    handler = logging.StreamHandler(sys.stdout)
+    handler.setFormatter(fmt)
+    log.addHandler(handler)
+
+    log.setLevel(logging.DEBUG)
+
+
+def add_event_hook(hook, event: Event = None):
+    """
+    Adds an event hook to be dispatched on an event.
+    ----------
+    :param hook:
+        The hook to register for the given event type.
+        If `event` parameter is left empty, then it will run when any event is dispatched.
+    :param event:
+        The event the hook belongs to. This will dispatch when that specific event is
+        dispatched.
+    """
+    if event is not None and not isinstance(event, Event):
+        raise TypeError('Event parameter is not of type Event or None')
+
+    if not callable(hook) or not inspect.iscoroutinefunction(hook):
```

Comment:
Missing parentheses ?
![image](https://user-images.githubusercontent.com/19573589/59381539-dcf38d00-8d5b-11e9-82fa-a1a63b872565.png)

Reply:
For which conditional?

Reply:
`if not callable(hook) or not inspect.iscoroutinefunction(hook):`

Reply:
Oh wat, there are 2 "not". nvm then